// my-admin/src/store/modules/user.js
import Vue from 'vue'
import { ACCESS_TOKEN } from '#/store/mutation-types'
import { welcome } from '#/utils/util'

const user = {
    state: {
        token: '',
        name: '',
        welcome: '',
        avatar: '',
        roles: [],
        info: {}
    },

    mutations: {
        SET_TOKEN: (state, token) => {
            state.token = token
        },
        SET_NAME: (state, { name, welcome }) => {
            state.name = name
            state.welcome = welcome
        },
        SET_AVATAR: (state, avatar) => {
            state.avatar = avatar
        },
        SET_ROLES: (state, roles) => {
            state.roles = roles
        },
        SET_INFO: (state, info) => {
            state.info = info
        }
    },

    actions: {
        // 登录
        Login({ commit }, userInfo) {
            return new Promise(async (resolve, reject) => {
                let $cloud = this._vm.$cloud;
                $cloud.login(userInfo).then(response => {
                    Vue.ls.set(ACCESS_TOKEN, response.token, 7 * 24 * 60 * 60 * 1000)
                    commit('SET_TOKEN', response.token)
                    resolve(response)
                }).catch(error => {
                    reject(error)
                })
            })
        },

        // 获取用户信息
        GetInfo({ commit }) {
            return new Promise(async (resolve, reject) => {
                let $cloud = this._vm.$cloud;

                let userInfo = await $cloud.users().get();
                //权限信息
                let roleInfo = await $cloud.roles().nameget(userInfo.role);

                //规则列表
                let rulesList = await $cloud.rules().all();



                //循环到处列表
                let rulesGroup = {}
                rulesList.forEach((value, key) => {
                    rulesGroup[value.name] = value;
                })

                //权限列表
                let permissionList = [];
                //权限内容
                let permissions = []


                //循环处理权限
                for (let key in roleInfo.rules) {
                    let value = roleInfo.rules[key];
                    //存储规则信息
                    let rulesInfo = {
                        "roleId": userInfo.role,
                        "permissionId": key,
                        "actionList": [],
                        "actionEntitySet": []
                    }

                    //存储两项数据
                    for (let _key in value) {
                        let _value = value[_key];
                        rulesInfo.actionList.push(rulesGroup[_value].expression);
                        rulesInfo.actionEntitySet.push(rulesGroup[_value])
                    }

                    //存储权限列表
                    permissions.push(rulesInfo);
                }

                //数据转换保存
                roleInfo.permissionList = permissionList
                roleInfo.permissions = permissions;
                userInfo.role = roleInfo;


                commit('SET_ROLES', userInfo.role)
                commit('SET_INFO', userInfo)

                commit('SET_NAME', { name: userInfo.name, welcome: welcome() })
                commit('SET_AVATAR', userInfo.avatar)

                resolve(userInfo)

            })
        },

        // 登出
        Logout({ commit, state }) {
            return new Promise((resolve) => {
                resolve()
                commit('SET_TOKEN', '')
                commit('SET_ROLES', [])
                Vue.ls.remove(ACCESS_TOKEN)
            })
        }

    }
}

export default user
