// my-admin/src/config/router.config.js
// eslint-disable-next-line
import { UserLayout, BasicLayout, RouteView, BlankLayout, PageView, PublicLayout } from '#/layouts'

/**
 * 基础路由
 * @type { *[] }
 */
export const constantRouterMap = [{
    path: '/public',
    component: UserLayout,
    redirect: '/public/login',
    hidden: true,
    children: [{
            path: 'login',
            name: 'login',
            component: () => import( /* webpackChunkName: "user" */ '#/views/public/Login')
        },
        {
            path: 'register',
            name: 'register',
            component: () => import( /* webpackChunkName: "user" */ '#/views/public/Register')
        },
        {
            path: 'register-result',
            name: 'registerResult',
            component: () => import( /* webpackChunkName: "user" */ '#/views/public/RegisterResult')
        },
        {
            path: 'AppInfo',
            name: 'AppInfo',
            component: () => import( /* webpackChunkName: "user" */ '#/views/public/AppInfo')
        },
        {
            path: 'recover',
            name: 'recover',
            component: undefined
        }
    ]
}, {
    path: '/404',
    component: () => import( /* webpackChunkName: "fail" */ '#/views/exception/404')
}]
