// my-admin/build/config.js
/*
 * @Author: sean
 * @Date:   2020-03-11 16:00:12
 * @Last Modified by:   wiki
 * @Last Modified time: 2020-05-11 16:16:49
 */
const path = require("path");

//常量
const { DS, DEBUG, FRAME_DIR, ROOT_DIR, ROOT_SRC, isProd } = require('./basics');

const webpack = require('webpack')
const { CleanWebpackPlugin } = require('clean-webpack-plugin')
const VueLoaderPlugin = require('vue-loader/lib/plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const createThemeColorReplacerPlugin = require('./plugin.config')
const openBrowserWebpackPlugin = require("open-browser-webpack-plugin"); // 打开浏览器
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const ProgressBarPlugin = require('progress-bar-webpack-plugin');
const package_c = require(ROOT_DIR + '/package.json')


console.log("process.env.NODE_ENV", process.env.NODE_ENV)

module.exports = {
    mode: 'development',
    devtool: 'cheap-module-eval-source-map',
    entry: ['webpack-hot-middleware/client?noInfo=true&reload=true', ROOT_SRC + '/main.js'],
    output: {
        filename: 'js/[name].[hash].js',
        publicPath: "/",
        path: path.resolve(ROOT_DIR, 'dist')
    },
    resolve: {
        extensions: ['.js', '.vue', '.jsx', '.json'],
        alias: {
            "#": ROOT_SRC,
            "@": path.resolve(ROOT_DIR, './')
        }
    },
    stats: {
        children: false,
    },
    module: {
        rules: [{
                test: /\.vue$/,
                use: ['vue-loader']
            },
            {
                test: /\.(png|jpe?g|gif|svg)(\?.*)?$/,
                loader: 'file-loader',
                include: [path.resolve(__dirname, "../")],
                options: {
                    esModule: false, // 这里设置为false
                    name: 'assets/[name].[ext]?[hash]',
                    limit: 102400
                }
            },
            {
                test: /\.css$/,
                use: [
                    MiniCssExtractPlugin.loader,
                    'css-loader',
                    'postcss-loader'
                ]
            },
            {
                test: /\.less$/,
                use: [{
                    loader: "style-loader"
                }, {
                    loader: "css-loader"
                }, {
                    loader: "less-loader",
                    options: {
                        javascriptEnabled: true
                    }
                }]
            },
            {
                test: /(\.js|jsx?)$/,
                include: [path.resolve(__dirname, "../")],
                use: ['babel-loader']
            }
        ]
    },
    plugins: [
        new webpack.DefinePlugin({
            'process.env': {
                NODE_ENV: '"development"',
                VUE_APP_PREVIEW: '"true"',
                CLOUD_VUE_DEV: package_c.CLOUD_VUE_DEV || false,
                CLOUD_VUE_KEY: `'${package_c.CLOUD_VUE_KEY}'` || "0000",
                CLOUD_VUE_HOST: `'${package_c.CLOUD_VUE_HOST}'` || false
            }
        }),
        new openBrowserWebpackPlugin({
            url: "http://localhost:8888"
        }),
        new ProgressBarPlugin(),
        new MiniCssExtractPlugin({
            filename: "css/[name].css",
            chunkFilename: "css/[id].css"
        }),
        createThemeColorReplacerPlugin(),
        new webpack.HotModuleReplacementPlugin(),
        new HtmlWebpackPlugin({
            template: path.resolve(ROOT_SRC, 'template.html'),
            // filename: '../index.html',
        }),
        new VueLoaderPlugin(),
        new CleanWebpackPlugin()
    ]
}
